File: typescript/src/ingestion.ts
import path from "node:path";
import { stat } from "node:fs/promises";
import { fileURLToPath } from "node:url";

import {
  discoverMarkdown,
  isReservedDocument,
  markdownFacts,
  OkfParserError,
  readUtf8Ordered,
  splitOptionalFrontmatter,
} from "./core.js";
import type { FrontmatterValue, LoadOptions, MarkdownFacts } from "./core.js";

export type IngestionCapability = "identity" | "body" | "frontmatter" | "markdown_facts";

export interface DocumentEnvelope {
  readonly ordinal: number;
  readonly path: string;
  readonly size: number | null;
  readonly classification: string;
  readonly body?: string;
  readonly frontmatter?: Readonly<Record<string, FrontmatterValue>> | null;
  readonly facts?: MarkdownFacts;
  readonly error?: string;
}

export interface IngestionOptions extends LoadOptions {
  readonly capabilities?: readonly IngestionCapability[];
}

function messageOf(error: unknown): string {
  return error instanceof Error ? error.message : String(error);
}

export async function* ingestDocuments(
  rootInput: string | URL,
  options: IngestionOptions = {},
): AsyncGenerator<DocumentEnvelope> {
  const root = path.resolve(rootInput instanceof URL ? fileURLToPath(rootInput) : rootInput);
  const requested = new Set<IngestionCapability>(options.capabilities ?? ["identity"]);
  requested.add("identity");
  const paths = await discoverMarkdown(rootInput, options);

  if (requested.size === 1) {
    for (const [ordinal, filePath] of paths.entries()) {
      options.signal?.throwIfAborted();
      const relative = path.relative(root, filePath).split(path.sep).join("/");
      let size: number | null = null;
      let error: string | undefined;
      try {
        size = (await stat(filePath)).size;
      } catch (reason) {
        error = messageOf(reason);
      }
      yield Object.freeze({
        ordinal,
        path: relative,
        size,
        classification: error === undefined
          ? isReservedDocument(filePath) ? "reserved" : "candidate"
          : "invalid",
        ...(error === undefined ? {} : { error }),
      });
    }
    return;
  }

  const concurrency = options.readConcurrency ?? 32;
  if (!Number.isSafeInteger(concurrency) || concurrency < 1 || concurrency > 256) {
    throw new OkfParserError(
      "OKF_READ_CONCURRENCY",
      "read concurrency must be an integer from 1 through 256",
    );
  }
  let ordinal = 0;
  for await (const loaded of readUtf8Ordered(paths, concurrency, options.signal)) {
    const relative = path.relative(root, loaded.filePath).split(path.sep).join("/");
    if ("error" in loaded) {
      yield Object.freeze({
        ordinal: ordinal++,
        path: relative,
        size: null,
        classification: "invalid",
        error: messageOf(loaded.error),
      });
      continue;
    }
    try {
      const [frontmatter, body] = splitOptionalFrontmatter(loaded.text);
      const reserved = isReservedDocument(loaded.filePath);
      const conceptType = frontmatter?.type;
      const classification = reserved
        ? "reserved"
        : typeof conceptType === "string" && conceptType !== ""
          ? `typed:${conceptType}`
          : "untyped";
      yield Object.freeze({
        ordinal: ordinal++,
        path: relative,
        size: Buffer.byteLength(loaded.text),
        classification,
        ...(requested.has("body") ? { body } : {}),
        ...(requested.has("frontmatter") ? { frontmatter } : {}),
        ...(requested.has("markdown_facts") ? { facts: markdownFacts(body) } : {}),
      });
    } catch (error) {
      yield Object.freeze({
        ordinal: ordinal++,
        path: relative,
        size: Buffer.byteLength(loaded.text),
        classification: "invalid",
        error: messageOf(error),
      });
    }
  }
}

File: src/okf_parser/ingestion.py
"""Capability-driven ordered ingestion for large Markdown corpora."""

from __future__ import annotations

from dataclasses import dataclass
from enum import StrEnum
from typing import TYPE_CHECKING

from okf_parser.discovery import discover_markdown
from okf_parser.exclusion import ExclusionRules
from okf_parser.parser import (
    DocumentParseError,
    MarkdownFacts,
    is_reserved_document,
    markdown_facts,
    split_optional_frontmatter,
)

if TYPE_CHECKING:
    from collections.abc import Iterator, Sequence
    from pathlib import Path

    from okf_parser.models import YamlValue


class IngestionCapability(StrEnum):
    """Per-document work requested by an ingestion consumer."""

    IDENTITY = "identity"
    BODY = "body"
    FRONTMATTER = "frontmatter"
    MARKDOWN_FACTS = "markdown_facts"


@dataclass(frozen=True, slots=True)
class DocumentEnvelope:
    """One deterministic ingestion result, projected to requested capabilities."""

    ordinal: int
    path: str
    size: int | None
    classification: str
    body: str | None = None
    frontmatter: dict[str, YamlValue] | None = None
    facts: MarkdownFacts | None = None
    error: str | None = None


def ingest_documents(
    root: Path,
    capabilities: Sequence[IngestionCapability] = (IngestionCapability.IDENTITY,),
    exclude: Sequence[str] = (),
) -> Iterator[DocumentEnvelope]:
    """Yield Markdown documents in deterministic path order with projection pushdown.

    An identity-only request performs discovery and stat calls but does not open or
    decode file contents. Other capabilities read each file once and reuse the same
    decoded text and frontmatter boundary result.
    """
    root = root.resolve()
    if not root.is_dir():
        msg = f"bundle root is not a directory: {root}"
        raise NotADirectoryError(msg)

    requested = frozenset(capabilities) | {IngestionCapability.IDENTITY}
    needs_text = requested != {IngestionCapability.IDENTITY}
    paths = discover_markdown(root, ExclusionRules.read(root, exclude))

    for ordinal, path in enumerate(paths):
        relative = path.relative_to(root).as_posix()
        classification = "reserved" if is_reserved_document(path) else "candidate"
        try:
            size = path.stat().st_size
            if not needs_text:
                yield DocumentEnvelope(ordinal, relative, size, classification)
                continue

            text = path.read_text(encoding="utf-8")
            frontmatter, body = split_optional_frontmatter(text)
            if classification != "reserved":
                concept_type = None if frontmatter is None else frontmatter.get("type")
                classification = (
                    f"typed:{concept_type}"
                    if isinstance(concept_type, str) and concept_type
                    else "untyped"
                )
            facts = (
                markdown_facts(body) if IngestionCapability.MARKDOWN_FACTS in requested else None
            )
            yield DocumentEnvelope(
                ordinal=ordinal,
                path=relative,
                size=size,
                classification=classification,
                body=body if IngestionCapability.BODY in requested else None,
                frontmatter=(frontmatter if IngestionCapability.FRONTMATTER in requested else None),
                facts=facts,
            )
        except (DocumentParseError, OSError, UnicodeDecodeError) as exc:
            yield DocumentEnvelope(
                ordinal=ordinal,
                path=relative,
                size=None,
                classification="invalid",
                error=str(exc),
            )
